doc/isconf.8
.TH "isconf(8)" 1 "02/24/2015" "ISconf 4.2.8.253"


.SH NAME

.P
isconf \- infrastructure build and configuration manager

.SH SYNOPSIS

.P
\fBisconf\fR [\fB\-Dhrq\fR] [\fB\-c\fR \fIconfig\fR] [\fB\-m\fR \fImessage\fR] \fIverb\fR [\fIverb_args\fR] ...

.SH QUICK START

.P
First, follow the short installation instructions in the INSTALL file
that came with this package.  It's best to do this on whatever you're
using as a golden master image, then deploy that image to all of your
machines.  If you're only setting up a few machines and have no image
server, then you \fImight\fR be able to get away with installing each
manually from the vendor CD, if you carefully install them each the
same way.

.P
Later, to install the latest version of package 'foo' on ten thousand
hosts, including any hosts that are currently down or not yet built,
you can log into any host and say this:

.nf
    cd /tmp
    wget http://example.com/foo-1.2.tar.gz  
    isconf start
    isconf lock just a comment about installing foo
    isconf snap foo-1.2.tar.gz  
    isconf exec tar -xzvf foo-1.2.tar.gz 
    isconf exec make -C foo-1.2 install
    isconf exec rm -rf foo-1.2.tar.gz foo-1.2
    isconf ci
.fi


.P
\&...then, on the other 9,999 hosts, run this during boot or from cron:

.nf
    isconf start
    isconf up
.fi


.P
If you're only managing a few machines, you can probably get away with
not starting the isconf daemon at boot \-\- as above, just start it
manually when you need it by saying 'isconf start'.  If you're turning
these machines over to someone else for long\-term support, don't want
to teach them isconf, and expect them to manually make a mess anyway,
then it would make sense to leave the daemon off when you're done
anyway.  See \fBBUGS/RESTRICTIONS\fR for some security reasons why it
also makes sense to leave the daemon off when you're done.

.SH DESCRIPTION

.P
See the GLOSSARY below for terms and concepts.

.P
ISconf can be thought of as a cross between \fBsudo\fR(8) and a
distributed version control tool like Git or Bitkeeper.  Changes you
make via ISconf are journaled and added to a distributed repository,
queuing them for execution on other target machines.  Those other
target machines do not need to be running, or even be built, at the
time you check in changes.  As you turn on, build, reboot, and/or run
\&'isconf up' on other machines, ISconf consults the journal and
executes the same changes, in the same order, on each machine.

.P
The ISconf architecture is completely peer\-to\-peer; there are no
central servers or other single points of failure, and it is designed
for use in partially\-partitioned networks such as DMZ environments.
The command\-line client talks to a daemon which runs on each
machine.  The daemon, usually started at boot, handles distributed
file storage, locking, and network communications.

.P
ISconf is not intended for use in environments where you want to make
manual, ad\-hoc, or other out\-of\-band changes to machines.  If you
don't have the will to rebuild all of your machines from scratch so
you know what's on their disks, don't care about disaster recovery,
don't need to keep any of your machines in lock\-step with each other,
don't need to test O/S changes before deploying them to production,
aren't as interested in O/S patch management, or still want to log in
as root on target machines and make arbitrary untracked changes, then
you don't want this package.  

.SH BACKGROUND

.P
One hundred years ago, automobiles were built by hand.  Each vehicle
was unique, composed of parts which were often crafted on the spot.
Repairs were expensive and frequent, owners needed to be mechanics,
fleetwide engineering changes were non\-existent.

.P
Then came mass production.  Today, a single automotive assembly line
produces vehicles of varying colors and options, all built from the
same basic design and tooling.  Replacement parts are interchangeable;
technicians bolt engineering changes onto existing vehicles with a
reasonable expectation that the parts will fit.  Economies of scale
have led to highly optimized designs, performance, and usability.
Drivers turn the key and go.

.P
Most IT departments are nowhere near that sort of capability; they
still install and maintain operating systems and applications by hand.
Each machine is unique, reliability is elusive, users become
technicians, fixes often require re\-engineering, most outages are
caused by other fixes, and infrastructure\-wide changes are fraught
with peril if they are possible at all.  Even basic security patches
are, as a rule, applied sporadically.

.P
ISconf provides some of the standardized tooling needed for
deterministic, reproducible management of UNIX machines \-\- the kind of
reproducibility you can count on for consistency, disaster recovery,
reliability, security, and auditability.  ISconf manages hosts over
their entire lifecycle following initial install, allowing you to
continue to test and deploy both major and minor changes well after
the target hosts have been placed into production service.  With this
tool you can safely replace kernels and bootloaders, install new
patches, packages, and tarballs, run arbitrary commands, and even
re\-install the entire operating system under program control, and do
it all in a way that can be consistently reproduced on other current
or future machines.

.P
Over the last decade, users of earlier versions of ISconf have found
that this consistency gives systems administrators enough breathing
room to "get ahead of the ticket curve", reclaim more of their nights
and weekends, and to, finally, begin to do more engineering and less
firefighting.

.SH PREREQUISITES

.P
To do deterministic and repeatable host management, there are some
things you need to do in addition to just installing and using ISconf.
Above all, you need to maintain a reasonable level of control over the
root\-owned bits which you place on your disks, both during initial
install as well as throughout their lifetime. 

.P
Automated systems administration is all about making self\-modifying
code behave consistently.  If you don't start from a known state and
keep it that way, then you can make no assertions about how your
machines will behave in comparison with each other \-\- a change which
works on one host may not work on others.  Once you've destroyed this
consistency, you can no longer count on QA, disaster recovery, load
balancers, distributed applications, HA clusters, new deployments, or
even single machine rebuilds to work correctly.

.RS
.IP 1. 3
If two or more hosts are supposed to act the same, then you need to
install them from the same disk image.  This applies to rebuilds of a
single host as well as multiple installs of identical hosts.  See
\fBbase image\fR in the glossary.

.IP 2. 3
Your host install tool needs to be able to capture an image of an
existing machine, save it on an install server, then dump that image
onto subsequent machines verbatim, altering only those things which
are supposed to be unique, such as IP address and hostname.  Among
Linux installers, for example, systemimager meets this requirement;
kickstart does not.  Under Solaris, you'll need to use Flash Archives,
not Jumpstart.  See \fBcheckpoint image\fR in the glossary.

.IP 3. 3
After initial install, you need to manage hosts exclusively with
ISconf \-\- no manual or other out\-of\-band changes.  There is one
semi\-exception to this rule:  You might want to use another tool to
manage environmentally\-influenced configuration files.  You'll want to
manage the binaries of that tool using ISconf, and take care to ensure
that the external tool manages only those files which it must.  See
\fBenvironmental data\fR in the glossary for more discussion.
.RE

.SH FLAGS

.P
Flags appear only after the isconf command name, not after
subcommands (as opposed to e.g. CVS).

.TP
\fB\-c\fR \fIconfig\fR  
Top\-level configuration file.  Defaults to /etc/is/main.cf.  

.TP
\fB\-D\fR
Show debug info on stderr.

.TP
\fB\-m\fR
Message \-\- human\-readable comment describing the change.  Required
only when locking.  This flag is deprecated and is likely to be
removed; see the 'lock' verb below for how to provide the message 
in a forward\-compatible way.

.TP
\fB\-r\fR
Allow reboot if needed.  Used only with the 'up' verb below.  Also
see the 'reboot' verb.
Ordinarily you would execute 'isconf \-r up' from an rc script,
which is a relatively safe time to allow reboots.  

This flag has no effect unless there is a 'reboot' operation
pending in the journal.  If there is a 'reboot' pending, then this
flag allows the reboot to take place.   You only want to provide
this flag at times when it's safe to reboot the local machine.

Without this flag, if 'isconf up' encounters a 'reboot' operation
during journal replay, the replay stops, an error message is
issued, and subsequent changes are not applied.  You'll need to
run 'isconf \-r up' to continue past this point \-\- we cannot assume
that the later changes will work without the reboot.

.TP
\fB\-q\fR
Quiet \-\- don't show verbose output.

.TP
\fB\-V\fR
Version \-\- show ISconf version.

.SH SUBCOMMANDS

.P
Subcommands are often called 'verbs' in ISconf documentation and
usage.  They can be grouped into the following categories:

.SS Changing disk state

.P
    lock, unlock, snap, exec, reboot, ci, up

.SS Branch management

.P
    fork, migrate

.SS Daemon management

.P
    start, stop, restart

.P
The following is a detailed description of all subcommands, in
alphabetical order.  In these descriptions, the \fBorigin\fR host is the
host where a user executes \fBlock\fR, \fBsnap\fR, \fBexec\fR, \fBreboot\fR,
or \fBci\fR, and the \fBtarget\fR host is where a user executes 
\fBup(date)\fR.

.TP
\fBci\fR
Check in local changes, such as \fBsnap\fR or \fBexec\fR, and release
branch lock.

Run on origin.

.TP
\fBexec\fR \fIcommand args ...\fR
Execute an arbitrary command.  Causes the command to be executed
immediately on the local machine, and queued for
execution on target machines after \fBci\fR.

Example:

.nf
    isconf lock "permanently shut down apache"
    isconf exec /etc/rc2.d/S85apache stop
    isconf exec rm /etc/rc2.d/S85apache 
    isconf ci
.fi



If you want to embed shell redirects or pipes in the \fBexec\fR
arguments, then you'll need to wrap the arguments in a shell
invocation.  For example, this *won't* do what you want \-\- it will
only change /etc/motd on the origin machine:

.nf
    isconf exec echo web server down > /etc/motd
.fi



Here's what you really want instead:

.nf
    isconf exec sh -c "echo web server down > /etc/motd"
.fi



.TP
\fBfork\fR \fInewbranch\fR
Create a new branch from the current branch, and migrate the local
host onto the new branch.  The
original branch is the "parent" branch, and the new branch is the
"child" branch.  

If host A executes a \fBfork\fR, then it is the only host moved to
the
branch; hosts B and C do not change.  If you want B or C to move
to the new branch as well, see \fBmigrate\fR.

Low\-level implementation:  Since a journal describes the details
of a branch, then a fork essentially just copies the entire
journal contents from the parent branch into a new journal named
after the child branch, then runs the \fBmigrate\fR code path.

.TP
\fBlock\fR \fImessage\fR ...
Lock the branch.  Required before \fBsnap\fR, \fBexec\fR, \fBreboot\fR,
or \fBci\fR, and recommended before \fBfork\fR and \fBmigrate\fR.  The
\fImessage\fR will be recorded in the journal for each subsequent
transaction until the next \fBci\fR.

.TP
\fBmigrate\fR \fIbranchname\fR 
Migrates the local host onto a new branch.  In human language
this means the host is going to change roles.  

Switching a host to a new branch is only possible if the new
branch is a child of the host's old branch, and if there have been
no transactions executed on the host since the new branch was
forked off \-\- in other words, the new branch's journal content
needs to be a contiguous superset of the old branch's journal
content.  If these conditions aren't met, \fBmigrate\fR will
exit with a non\-zero return code.

.TP
\fBreboot\fR 
Reboots the machine.  Before reboot, adds a journal entry which
will cause all target machines on this branch to reboot at the same
point in their build.   For example, this is what you might do to
install and boot a new kernel:

.nf
    isconf lock "upgrade to 2.6.20"
    isconf snap kernel-2.6.20-1.i686.rpm
    isconf exec rpm -ivh kernel-2.6.20-1.i686.rpm
    isconf reboot
    isconf ci

    # on other machines
    isconf -r up
.fi



Apply thought when using this verb; 'isconf up' (without the \-r)
won't finish if there is a 'reboot' pending as the next action in
the journal.  You need 'isconf \-r up' \-\- and you don't want to put
that in crontab, unless you really don't mind your
machines rebooting at that time.  See the \fB\-r\fR flag for details.

Never say 'isconf exec reboot' \-\- that will only reboot the local
machine, and will never create any sort of journal entry; the
reboot kills isconf itself before the journal entry can be made.
Always say 'isconf reboot' instead.

By default, ISconf runs 'shutdown \-r now' to cause the reboot.  If
you want or need to use a different command, see the IS_REBOOT_CMD
environment variable below.

.TP
\fBrestart\fR
Restart the daemon.  Equivalent to a \fBstop\fR followed by a
\fBstart\fR.

.TP
\fBsnap\fR \fIfilename\fR

Snapshot a file for install on target machines.  Preserves the current 
contents, permissions, and mode bits of the file.
After \fBci\fR, any target host on the same branch can
run 'isconf up', which will cause ISconf to install the file on 
the target host.

.TP
\fBstart\fR
.TP
\fBstop\fR
Start or stop the daemon.

.TP
\fBunlock\fR 
Break the lock on the local branch.  
Use with great care.  
This reverses the effect of a 
\fBlock\fR, invalidates the work stored in \fBjournal.wip\fR on
the locking machine, and will likely require the person who set
the lock to discard their work and/or rebuild the machine where
the lock was made.  

Generally speaking, it's better to pick up the telephone and call
the person who set the lock, asking them politely to finish
whatever they were doing and check it in, rather than use this
subcommand.

.TP
\fBup\fR 
Update.  Causes the isconf daemon to attempt execution of any new 
transactions in the
journal.  Errors and messages are copied to stderr and stdout of
\fBisconf\fR as well as to syslog.  Exits with a non\-zero return
code in case of error.

If used with \fB\-r\fR, and if a pending \fBreboot\fR entry is 
encountered in the journal, then the host will reboot.

.SH ENVIRONMENT

.P
ISconf behavior is controlled predominantly by environment
variables.  These can be set and exported before starting or
restarting the isconf daemon, or can be set in configuration
files, usually \fBmain.cf\fR.  Any
variables set in the environment will be overridden by those set
in the configuration file.

.TP
\fBIS_DOMAIN\fR 
ISconf domain name \-\- more or less equivalent to an AFS cell name
or a Kerberos realm name; all of the machines sharing this name
will share in the distributed cache that makes up the ISconf
repository.  Normally you'd want all of the machines in a given
legal entity \-\- the same corporation, for instance, to use the
same domain name.   This is an arbitrary string, but by convention it
is usually based on the DNS domain name.

Rather than set this in an environment variable, you're better off
populating the \fB/var/is/conf/domain\fR file, below.

See the \fBdomain\fR glossary entry.

.TP
\fBIS_HOME\fR 
The base directory which ISconf uses for data storage.  Defaults
to \fB/var/is\fR.

.TP
\fBIS_HMAC_KEYS\fR 
The name of a file which contains a list of HMAC keys.
See the \fBhmac_keys\fR file below.

.TP
\fBIS_HTTP_PORT\fR 
The port number which each ISconf HTTP server listens on.  Used only for
file fetches between machines, and is likely to be deprecated in a
near\-future release.  Defaults to port 65028.

.TP
\fBIS_NETS\fR 
The name of a file which contains a list of broadcast and/or host
addresses which ISconf should advertize file updates to.  See
\fBnets\fR file below.  Likely to change in a future release.

.TP
\fBIS_NOBROADCAST\fR 
Boolean.  If set, do not send UDP broadcast packets; only send
UDP point\-to\-point packets to the addresses listed in **nets*
file.  Likely to change in a future release.

.TP
\fBIS_PORT\fR 
The port number which ISconf daemons use to communicate between each
other.  Right now this is UDP only, but TCP will be added in
4.2.7, and UDP is likely to be deprecated.  Defaults to port 65027.

.TP
\fBIS_REBOOT_CMD\fR 
The command which ISconf uses to reboot the machine in response to
an 'isconf reboot' request.  Defaults to "shutdown \-r now".

.SH FILES

.TP
\fB/etc/is/main.cf\fR
Top\-level configuration file for ISconf.  See CONFIGURATION for
details.  As of this writing, ISconf does not distribute this file for
you.  In earlier versions, we used to simply rsync it from a
central server at the beginning of each execution.  In a near\-future 
version, look for it to be managed by the distributed cache.

.TP
\fB/var/is\fR
See \fBIS_HOME\fR above.

.TP
\fB/var/is/conf/domain\fR
Single\-line file, newline optional, containing only the string
which is to be used for the ISconf domain name.  See \fBIS_DOMAIN\fR
above. 

.TP
\fBhmac_keys\fR

HMAC key list, one key per line.  See \fBIS_HMAC_KEYS\fR.  If this
file exists and contains properly\-formatted keys, then RFC 2104 HMAC
authentication is enabled; wire messages which are not properly
authenticated will be ignored.  

The first key in the list is used for generating authentication
codes on all outgoing messages, and is the first key tried when
authenticating inbound messages.  If the first key fails to
authenticate an inbound message, and if more than one key is
listed in the file, then the second and subsequent keys are tried,
in order.  This mechanism enables you to update the primary key 
while preserving backward compatibility with older keys, allowing
for a transition period.  

When updating keys, it's a good idea to first add the new key as a
secondary key to the hmac_keys file, and deploy that to all
machines.  Once you're sure that \fBall\fR of your machines (and
install images) have the new key, then move the new key up to the
primary position in the file, leaving any old key(s) in the file
as secondaries, then deploy that.  Finally, once you're again sure
that \fBall\fR of your machines (and install images) are using the
new primary key, then (and only then) should you think about
retiring any old key(s).

Take care when deploying this file for the first time on hosts
which are already running ISconf; those ISconf daemons which get
it first will refuse to listen to any which don't yet have the
file; this will prevent further deployment if you're using ISconf
to deploy the file.  To prevent this from happening, you can
include the special key \fB+ANY+\fR at the end of the file.  If
encountered in the file, this special key disables HMAC
authentication of received messages, but does not prevent
generation of authentication codes on transmitted messages.  What
you want to do is deploy the file with one or more real keys
listed in it, followed by the \fB+ANY+\fR key.  The file might look
like this when first deployed:

.nf
    someauthenticationkey
    +ANY+

.fi



As you deploy the above file, hosts will begin sending
authenticated messages to each other using the
\fBsomeauthenticationkey\fR key, but will ignore the authentication
codes they receive.  Once you are sure that all of your hosts have
that copy of the file, then deploy the file again, this time with
the \fB+ANY+\fR key removed.  This will cause hosts to begin
checking received authentication codes against
\fBsomeauthenticationkey\fR, while discarding any messages not
properly authenticated.

For best security, each key should be about 20 bytes long; see RFC
2104.  Keys can can include any ASCII character except space,
newline, or the pound (hash) (#) sign.  Lines beginning with pound
signs are comments.  Blank lines are ignored.  If no keys are
found in the file, then the entire file is ignored, and HMAC
authentication is disabled.

ISconf checks for new versions of this file every 10 seconds when
it is processing inbound packets \-\- there is no need to restart
the ISconf daemon.

The hash function used internally is SHA\-1, with Python's
\fBhmac\fR module doing the real work.

You should ensure that this file is only readable by root.

This entire mechanism is likely to change and/or be replaced by
PGP key signatures in a future release.

.TP
\fBnets\fR
Network broadcast list \-\- see \fBIS_NETS\fR above.  See t/nets for
an example.  Likely to change.

.SH CONFIGURATION

.P
ISconf uses environment variables for its configuration, and these
variables are in turn passed on to any executables ISconf calls \-\- see
ENVIRONMENT.   These environment variables can be set in
/etc/is/main.cf.  The format of this file is similar to a makefile,
but whitespace is whitespace \-\- tabs aren't required.  Each stanza
looks like this:

.nf
    target: optional includes
        var1 = value
        var2 = value
.fi


.P
The 'target' string above is matched against the hostname; case is
significant.  If it contains dots, it's matched against the FQDN.  If
it starts with a caret (^) it is a regex matched against the FQDN.
The first matching target is the only one used, however the special
target named 'DEFAULT' is always matched.  Variables set in DEFAULT,
earlier includes, or earlier in the same stanza are overridden by
identically\-named variables which appear later in matched stanzas.
Comments are any text following a hash (#) on any line.

.P
You can see the resulting environment by using the \fB\-D\fR flag.

.P
Here's an example /etc/is/main.cf:

.nf
    DEFAULT:
        NTPSERVERS = ntp1 ntp2 bigben.ucsd.edu mcs.anl.gov
        IS_NETS=/etc/is/nets

    NET1:
        GATEWAY = 10.10.1.1

    NET2:
        GATEWAY = 10.10.2.1

    # The host 'scotty' will end up with these environment variables
    # set during the ISconf run:

lib/python/isconf/ISconf.py
# vim:set expandtab:
# vim:set foldmethod=indent:
# vim:set shiftwidth=4:
# vim:set tabstop=4:

from __future__ import generators
import ConfigParser
import copy
import email.Message
import email.Parser
import inspect
import md5
import os
import popen2
import random
import re
import select
import sha
import shutil
import socket
import sys
import time
import isconf
from isconf.Errno import iserrno
from isconf.Globals import *
from isconf import ISFS
from isconf.Kernel import kernel, Bus
from isconf.fbp822 import fbp822, Error822
from isconf.Socket import Timeout

class CLIServerFactory:

    def __init__(self,socks):
        self.socks = socks

    def run(self):
        while True:
            slist = []
            yield self.socks.rx(slist)
            for sock in slist:
                server = CLIServer(sock=sock)
                kernel.spawn(server.run())

class CLIServer:

    def __init__(self,sock):
        self.transport=sock
        self.verbose = False
        self.debug = False

    def run(self):
        yield kernel.sigbusy 
        debug("CLIServer running")
        fbp = fbp822()

        # set up FBP buses
        tocli = Bus('tocli')

        # process messages from client
        proc = kernel.spawn(self.process(transport=self.transport,outpin=tocli))
        # send messages to client
        res = kernel.spawn(self.respond(transport=self.transport,inpin=tocli))
        # merge in log messages
        log = kernel.spawn(self.merge(tocli,BUS.log))

        # heartbeat to client
        kernel.spawn(self.heartbeat(transport=self.transport))

        # wait for everything to quiesce
        yield kernel.siguntil, proc.isdone
        yield kernel.sigsleep, .1
        while True:
            yield None
            i=0
            for q in (tocli,BUS.log):
                if q.busy():
                    i+=1
                    continue
            if i == 0: 
                break

        debug("telling client to exit")
        yield kernel.sigsleep, .1 # XXX 
        tocli.tx(fbp.mkmsg('rc',0))
        # tocli.close()

    def merge(self,outbus,inbus):
        while True:
            mlist = []
            yield inbus.rx(mlist)
            for msg in mlist:
                if msg in (kernel.eagain,None):
                    continue
                if outbus.state == 'down':
                    return
                if msg is kernel.eof:
                    return
                outbus.tx(msg)

    def process(self,transport,outpin):
        while True:
            yield None
            # read messages from client
            mlist = FBP.fromFile(stream=transport)
            for msg in mlist:
                if msg in (kernel.eagain,None):
                    continue
                if outpin.state == 'down':
                    return
                if msg is kernel.eof:
                    # outpin.close()
                    return
                debug("from client:", str(msg))
                rectype = msg.type()
                if rectype != 'cmd':
                    error(iserrno.EINVAL, 
                        "first message must be cmd, got %s" % rectype)
                    return
                self.verbose = msg.head.verbose
                self.debug = msg.head.debug
                data = msg.payload()
                opt = dict(msg.items())
                if opt['message'] == 'None':
                    opt['message'] = None
                # XXX why don't we just pass the whole message to Ops()?
                opt['reboot_ok'] = msg.head.reboot_ok
                debug(opt)
                verb = msg['verb']
                debug("verb in process",verb)
                if verb != 'lock' and opt['message'] != None:
                    error(iserrno.EINVAL, "-m is only valid on lock")
                if verb == 'exec': verb = 'Exec' # sigh
                if verb == 'shutdown': 
                    kernel.shutdown()
                    return
                args=[]
                if len(data):
                    data = data.strip()
                    args = data.split('\n')
                ops = Ops(opt=opt,args=args,data=data,outpin=outpin)
                try:
                    func = getattr(ops,verb)
                except AttributeError:
                    error(outpin,iserrno.EINVAL,verb)
                    return
                # start command processor, wait for it to finish
                debug("process calling",verb)
                yield kernel.wait(func())
                return # XXX can't handle stdin yet

    def heartbeat(self,transport):
        msg = FBP.msg('heartbeat')
        while True:
            yield kernel.sigsleep,1
            if transport.state == 'down':
                return
            transport.write(str(msg))

    def respond(self,transport,inpin):
        while True:
            mlist = []
            yield inpin.rx(mlist)
            for msg in mlist:
                # if not hasattr(msg,'type'):
                if transport.state == 'down':
                    return
                if msg in (kernel.eagain,None):
                    continue
                if msg is kernel.eof:
                    # transport.close()
                    return
                # no logging in here!  causes a message loop...
                # debug("to client:", str(msg))
                rectype = msg.type()
                if rectype == 'debug' and not self.debug:
                    continue
                transport.write(str(msg))
                if msg.type() == 'rc':
                    transport.close()
                    return


class Ops:
    """ISconf server-side operations
    
    Each of these tasks *must* continue running until their operations
    are complete.  They or their called routines can send a non-zero rc
    message to outpin.  If they don't, process() will follow up with a
    zero rc.  Whichever rc message arrives at respond() first wins.
    
    """

    def __init__(self,opt,args,data,outpin):
        self.opt = opt
        self.args = args
        self.data = data
        self.outpin = outpin
        self.histfile = "%s/conf/history" % os.environ['IS_HOME']

        self.volname = branch()
        self.logname = self.opt['logname']
        self.volume = ISFS.Volume(
                self.volname,logname=self.logname,outpin=outpin,
                histfile=self.histfile)
        info("%s is on %s branch" % (os.environ['HOSTNAME'],branch()) )

    def ci(self):
        yield kernel.wait(self.volume.ci())

    def Exec(self):
        yield None
        cwd = self.opt['cwd']
        if not len(self.data):
            error(iserrno.EINVAL,"missing exec command")
            return
        yield kernel.wait(self.volume.Exec(self.data,cwd))

    def fork(self,migrate=False):
        debug("fork args",self.args)
        if self.volume.wip(): 
            error("local changes not checked in")
            return 
        if len(self.args) < 1:
            error("missing new branch name")
            return
        newbranch = self.args[0]
        debug("newbranch",newbranch)
        if migrate and newbranch == self.volname:
            error("%s is already on %s branch" % (
                    os.environ['HOSTNAME'], newbranch)
                    )
            return
        newvolume = ISFS.Volume(
                newbranch,logname=self.logname,outpin=self.outpin,
                histfile=self.histfile)
        yield kernel.wait(self.volume.pull(bg=True))
        yield kernel.wait(newvolume.pull(bg=True))
        if not migrate:
            if not newvolume.journal.copy(self.volume.journal):
                error("%s already exists -- did you mean 'migrate'?" % newbranch)
                return
            info("branch %s created" % (newbranch))
        else:
            if not self.volume.journal.migrate(newvolume.journal):
                error("migration of this host to %s failed: conflicting changes: %s is not a superset of %s" % (newbranch,newbranch,self.volname))
                return
        branch(newbranch)
        self.volname = newbranch
        self.volume = newvolume
        info("%s migrated to %s" % (os.environ['HOSTNAME'], newbranch))
        
    def lock(self):
        message = ''
        if self.opt['message']:
            message = self.opt['message']
        if self.args:
            message = "%s %s" % (message,' '.join(self.args))
        message = message.strip()
        if not len(message):
            error(iserrno.NEEDMSG,'did not lock %s' % self.volname)
            return
        yield kernel.wait(self.volume.lock(message))

    def migrate(self):
        yield None
        debug("migrate calling fork")
        yield kernel.wait(self.fork(migrate=True))

    def reboot(self):
        yield None
        debug("calling reboot")
        yield kernel.wait(self.volume.reboot())

    def snap(self):
        debug("starting snap")
        yield None
        if not len(self.args):
            error(iserrno.EINVAL,"missing snapshot pathname")
            return
        if len(self.args) > 1:
            error(iserrno.EINVAL,
                    "can only snapshot one file at a time (for now)")
            return
        path = self.args[0]
        cwd = self.opt['cwd']
        path = os.path.join(cwd,path)
        if not os.path.exists(path):
            error(iserrno.ENOENT,path)
            return
        if not os.path.isfile(path):
            error(iserrno.EINVAL,"%s is not a file" % path)
            return
        st = os.stat(path)
        src = open(path,'r')
        debug("calling open")
        dst = self.volume.open(path,'w')
        if not dst:
            return
        dst.setstat(st)
        while True:
            yield kernel.sigbusy
            data = src.read(1024 * 1024 * 1)
            if not len(data):
                break
            dst.write(data)
        src.close()
        debug("calling close")
        yield kernel.wait(dst.close())

    def unlock(self):
        yield None
        locker = self.volume.lockedby()
        self.volume.unlock()
        info("broke %s lock -- please notify %s" % (self.volname,locker))

    def up(self):
        reboot_ok = bool(self.opt.get('reboot_ok',False))
        if reboot_ok:
            debug("reboot_ok", repr(self.opt['reboot_ok']))
            info("may reboot...")
        yield kernel.wait(self.volume.update(reboot_ok=reboot_ok))

            
def branch(val=None):
    ishome = os.environ['IS_HOME']
    fname = "%s/conf/branch" % ishome
    if not os.path.exists(fname):
        val = 'generic'
    if val is not None:
        open(fname,'w').write(val+"\n")
    val = open(fname,'r').read().strip()
    return val

def XXXbusexit(errpin,code,msg=''):
    desc = iserrno.strerror(code)
    if str or msg:
        msg = "%s: %s\n" % (str(msg), desc)
    fbp=fbp822()
    if msg and code:
        warn("busexit: ", msg)
        msg = "isconf: error: " + msg
        errpin.tx(fbp.mkmsg('stderr',msg))
    errpin.tx(fbp.mkmsg('rc',code))
    # errpin.close()

def client(transport,argv,kwopt):

    """
    A unix-domain client of an isconf server.  This client is very
    thin -- all the smarts are on the server side.

    argv is e.g. ('snap', '/tmp/foo') 
    """

    # XXX convert to use global log funcs
    def clierr(code,msg=''):
        desc = iserrno.strerror(code)
        if len(msg):
            msg = "%s: %s" % (msg, desc)
        else:
            msg = desc.strip()
        warn("clierr: ", msg)
        return code

    fbp = fbp822()
    verb = argv.pop(0)
    if len(argv):
        payload = "\n".join(argv) + "\n"
    else:
        payload = ''
    logname = os.environ['LOGNAME']
    cwd = os.getcwd()
    msg = fbp.mkmsg('cmd',payload,verb=verb,logname=logname,cwd=cwd,**kwopt)

    # this is a blocking write...
    # XXX what happens here if daemon dies?
    debug("sending cmd")
    transport.write(str(msg))
    debug("cmd sent")

    # we should get a heartbeat message from server every few seconds  
    # XXX make this smaller, fix test case, test
    timeout = os.environ.get('IS_TIMEOUT_CLI',60)
    transport.timeout = timeout

    # sockfile = transport.sock.makefile('r')
    # stream = fbp.fromFile(sockfile,intask=False)
    stream = fbp.fromStream(transport,intask=False)
    # process one message each time through loop
    while True:
        debug("client loop")
        time.sleep(.1)
        try:
            msg = stream.next()
        except StopIteration:
            return clierr(iserrno.ECONNRESET)
        except Error822, e:
            return clierr(iserrno.EBADMSG,e)
        except Timeout:
            # XXX change this after logging cleaned up
            error("server timed out -- check /tmp/isconf.stderr and isconf.log")
        if msg in (kernel.eagain,None,kernel.sigbusy):
            continue
        rectype = msg.type()
        data = msg.payload()
        if rectype == 'heartbeat':
            debug("heartbeat from server")
        elif rectype == 'rc':
            code = int(data)
            return code
        elif rectype == 'stdout': 
            sys.stdout.write(data)
            sys.stdout.flush()
        elif rectype == 'stderr': 
            sys.stderr.write(data)
            sys.stderr.flush()
        elif rectype == 'reqstdin':
            for line in sys.stdin:
                msg = fbp.mkmsg('stdin',line)
                transport.write(str(msg))
            transport.shutdown()
        elif rectype == 'debug':
            debug(data)
        elif rectype == 'info':
            info(data)
        elif rectype == 'warn':
            warn(data)
        elif rectype == 'error':
            error(msg.head.rc,data)
        elif hasattr(os.environ,'DEBUG'):
            debug(str(msg))
        
